core/js/src/typespecs/prompt.ts

```diff
@@ -0,0 +1,122 @@
+import { z } from "zod";
+
+export const roleSchema = z.enum([
+  "system",
+  "user",
+  "assistant",
+  "function",
+  "tool",
+  "model",
+]);
+export type Role = z.infer<typeof roleSchema>;
+
+export const functionCallSchema = z.object({
+  name: z.string(),
+  arguments: z.string(),
+});
+
+const toolCallSchema = z.object({
+  id: z.string(),
+  function: z.object({
+    arguments: z.string(),
+    name: z.string(),
+  }),
+  type: z.literal("function"),
+});
+
+export const messageSchema = z.object({
+  content: z.string().default(""),
+  role: roleSchema,
+  name: z.string().optional(),
+  function_call: z.union([z.string(), functionCallSchema]).optional(),
+  tool_calls: z.array(toolCallSchema).optional(),
+});
+
+export const promptBlockDataSchema = z.union([
+  z.object({
+    type: z.literal("completion"),
+    content: z.string().default(""),
+  }),
+  z.object({
+    type: z.literal("chat"),
+    messages: z.array(messageSchema),
+    tools: z.string().optional(),
+  }),
+]);
+
+export type PromptBlockData = z.infer<typeof promptBlockDataSchema>;
+
+const braintrustModelParamsSchema = z.object({
+  use_cache: z.boolean().optional(),
+});
+
+const openAIModelParamsSchema = z.object({
+  temperature: z.number(),
+  top_p: z.number().optional(),
+  max_tokens: z.number().optional(),
+  frequency_penalty: z.number().optional(),
+  presence_penalty: z.number().optional(),
+  response_format: z
+    .union([z.literal(null), z.object({ type: z.literal("json_object") })])
+    .optional(),
+  tool_choice: z
+    .union([
+      z.literal("auto"),
+      z.literal("none"),
+      z.object({
+        type: z.literal("function"),
+        function: z.object({ name: z.string() }),
+      }),
+    ])
+    .optional(),
+});
+
+const anthropicModelParamsSchema = z.object({
+  max_tokens: z.number(),
+  temperature: z.number(),
+  top_p: z.number().optional(),
+  top_k: z.number().optional(),
+  max_tokens_to_sample: z
+    .number()
+    .optional()
+    .describe("This is a legacy parameter that should not be used."),
+});
+
+const googleModelParamsSchema = z.object({
+  temperature: z.number(),
+  maxOutputTokens: z.number().optional(),
+  topP: z.number().optional(),
+  topK: z.number().optional(),
+});
+
+const jsCompletionParamsSchema = z.object({});
+export const modelParamsSchema = braintrustModelParamsSchema.and(
+  z.union([
+    openAIModelParamsSchema,
+    anthropicModelParamsSchema,
+    googleModelParamsSchema,
+    jsCompletionParamsSchema,
+  ])
+);
+
+export type ModelParams = z.infer<typeof modelParamsSchema>;
+export const promptOptionsSchema = z.object({
+  model: z.string().optional(),
+  params: modelParamsSchema.optional(),
+  position: z.string().optional(),
+});
+
+export type PromptOptions = z.infer<typeof promptOptionsSchema>;
+
+export const promptDataSchema = z.object({
```

Comment:
For large objects that may get used in multiple places in an openAPI spec, I like to add a `.openapi("[TypeName]")` annotation, so that the generated spec can define a JSON schema type just for this object, and reference it in other places. Otherwise it'll get copied throughout everywhere it's used.